packages/SwingSet/src/kernel/vatManager/vat-warehouse.js

```diff
@@ -0,0 +1,275 @@
+    // NOTE: OPTIMIZATION OPPORTUNITY: replay vats in parallel
+
+    // instantiate all static vats
+    for (const [name, vatID] of kernelKeeper.getStaticVats()) {
+      logStartup(`allocateVatKeeper for vat ${name} as vat ${vatID}`);
+      // eslint-disable-next-line no-await-in-loop
+      await ensureVatOnline(vatID, recreate);
+    }
+
+    // instantiate all dynamic vats
+    for (const vatID of kernelKeeper.getDynamicVats()) {
+      logStartup(`allocateVatKeeper for dynamic vat ${vatID}`);
+      // eslint-disable-next-line no-await-in-loop
+      await ensureVatOnline(vatID, recreate);
+    }
+  }
+
+  /**
+   * @param { string } vatID
+   * @returns {{ enablePipelining?: boolean }
+   *  | void // undefined if the vat is dead or never initialized
+   * }
+   */
+  function lookup(vatID) {
+    const liveInfo = ephemeral.vats.get(vatID);
+    if (liveInfo) {
+      const { enablePipelining } = liveInfo;
+      return { enablePipelining };
+    }
+    const vatKeeper = kernelKeeper.getVatKeeper(vatID);
+    if (vatKeeper) {
+      const {
+        options: { enablePipelining },
+      } = vatKeeper.getSourceAndOptions();
+      return { enablePipelining };
+    }
+    return undefined;
+  }
+
+  /**
+   *
+   * @param {string} vatID
+   * @param {boolean=} makeSnapshot
+   * @returns { Promise<unknown> }
+   */
+  async function evict(vatID, makeSnapshot = false) {
+    assert(!makeSnapshot, 'not implemented');
+    assert(lookup(vatID));
+    const info = ephemeral.vats.get(vatID);
+    if (!info) return undefined;
+    ephemeral.vats.delete(vatID);
+    xlate.delete(vatID);
+    const vatKeeper = kernelKeeper.getVatKeeper(vatID);
+    if (vatKeeper) {
+      // not terminated
+      vatKeeper.closeTranscript();
+    }
+
+    // console.log('evict: shutting down', vatID);
+    return info.manager.shutdown();
+  }
+
+  /** @type { string[] } */
+  const recent = [];
+
+  /**
+   * Simple fixed-size LRU cache policy
+   *
+   * TODO: policy input: did a vat get a message? how long ago?
+   * "important" vat option?
+   * options: pay $/block to keep in RAM - advisory; not consensus
+   * creation arg: # of vats to keep in RAM (LRU 10~50~100)
+   *
+   * @param {string} currentVatID
+   */
+  async function applyAvailabilityPolicy(currentVatID) {
+    // console.log('applyAvailabilityPolicy', currentVatID, recent);
+    const pos = recent.indexOf(currentVatID);
+    // console.debug('applyAvailabilityPolicy', { currentVatID, recent, pos });
+    // already most recently used
+    if (pos + 1 === maxVatsOnline) return;
+    if (pos >= 0) recent.splice(pos, 1);
+    recent.push(currentVatID);
+    // not yet full
+    if (recent.length <= maxVatsOnline) return;
+    const [lru] = recent.splice(0, 1);
+    // console.debug('evicting', { lru });
+    await evict(lru);
+  }
+
+  /** @type {(vatID: string, d: VatDeliveryObject) => Promise<Tagged> } */
+  async function deliverToVat(vatID, delivery) {
+    await applyAvailabilityPolicy(vatID);
```

Comment:
I noted this in #2422

Reply:
filed #3279 to cover this, upon further analysis I don't think it would cause incorrect behavior, merely inefficiency, but I wrote up a plausible solution in that ticket